java/semantickernel-api/src/main/java/com/microsoft/semantickernel/FunctionInvocation.java

```diff
@@ -0,0 +1,191 @@
+package com.microsoft.semantickernel;
+
+import com.microsoft.semantickernel.exceptions.SKException;
+import com.microsoft.semantickernel.hooks.KernelHooks;
+import com.microsoft.semantickernel.hooks.KernelHooks.UnmodifiableKernelHooks;
+import com.microsoft.semantickernel.orchestration.FunctionResult;
+import com.microsoft.semantickernel.orchestration.InvocationContext;
+import com.microsoft.semantickernel.orchestration.KernelFunction;
+import com.microsoft.semantickernel.orchestration.KernelFunctionArguments;
+import com.microsoft.semantickernel.orchestration.PromptExecutionSettings;
+import com.microsoft.semantickernel.orchestration.ToolCallBehavior;
+import com.microsoft.semantickernel.orchestration.ToolCallBehavior.UnmodifiableToolCallBehavior;
+import com.microsoft.semantickernel.orchestration.contextvariables.ContextVariable;
+import com.microsoft.semantickernel.orchestration.contextvariables.ContextVariableType;
+import java.util.function.BiConsumer;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.CoreSubscriber;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.SynchronousSink;
+
+public class FunctionInvocation<T> extends Mono<FunctionResult<T>> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FunctionInvocation.class);
+
+    private final KernelFunction<?> function;
+    private final Kernel kernel;
+    @Nullable
+    private KernelFunctionArguments arguments;
+    @Nullable
+    private ContextVariableType<T> resultType;
+    @Nullable
+    private UnmodifiableKernelHooks hooks;
+    @Nullable
+    private PromptExecutionSettings promptExecutionSettings;
+    @Nullable
+    private UnmodifiableToolCallBehavior toolCallBehavior;
+
+
+    public FunctionInvocation(
+        Kernel kernel,
+        KernelFunction<T> function) {
+        this.function = function;
+        this.kernel = kernel;
+    }
+
+    public FunctionInvocation(
+        Kernel kernel,
+        KernelFunction<?> function,
+        @Nullable
+        ContextVariableType<T> resultType) {
+        this.function = function;
+        this.kernel = kernel;
+        this.resultType = resultType;
+    }
+
+    public FunctionInvocation<T> withArguments(
+        @Nullable
+        KernelFunctionArguments arguments) {
+        if (arguments == null) {
+            this.arguments = null;
+        } else {
+            this.arguments = new KernelFunctionArguments(arguments);
+        }
+        return this;
+    }
+
+    public <U> FunctionInvocation<U> withResultType(ContextVariableType<U> resultType) {
+        return new FunctionInvocation<>(
```

Comment:
I have changed the resultType to being immutable. The reason a new FunctionInvocation is created is because providing a resultType can change the type, notice that this function converts from `FunctionInvocation<T>` to `FunctionInvocation <U>` if the user requests a different type